app/domain/authentication/authn_oidc/authenticator.rb

```diff
@@ -0,0 +1,84 @@
+module Authentication
+  module AuthnOidc
+    class AuthenticationError < RuntimeError; end
+    class NotFoundError < RuntimeError; end
+    class OIDCConfigurationError < RuntimeError; end
+    class OIDCAuthenticationError < RuntimeError; end
+
+    class Authenticator
+
+      def initialize(env:)
+        @env = env
+      end
+
+      def valid?(input)
+        @authenticator_name = input.authenticator_name
+        @service_id = input.service_id
+        @conjur_account = input.account
+        @request_body = input.request.body.read
+
+        verify_service_enabled
+
+        oidc_authn_service = AuthenticationService.new(service.identifier, conjur_account)
+        id_token, user_info = oidc_authn_service.get_user_details(request_body)
+
+        # TODO: validate id_token - if not raise error
+
+        validate_user_info(user_info, id_token.sub)
+
+        username = user_info.preferred_username
+        input.instance_variable_set(:@username, username)
+
+        true
+      end
+
+      private
+
+      def authenticator_name
+        @authenticator_name
+      end
+
+      def conjur_account
+        @conjur_account
+      end
+
+      def service_id
+        @service_id
+      end
+
+      def request_body
+        @request_body
+      end
+
+      def service
+        @service ||= Resource["#{conjur_account}:webservice:conjur/#{authenticator_name}/#{service_id}"]
+      end
```

Comment:
`Resource` should be an injected dependency.

Reply:
Do you mean this - https://github.com/cyberark/conjur/blob/97f1e3a3b4edb9e7bc1f3da973f651ea9006b009/app/domain/authentication/webservice.rb#L25?

In the method below (`resource_id`) you d othe exact string building we're doing so I'm a bit confused.

Regarding the injected dependency - this will be done after the split.

Reply:
> In the method below (resource_id) you d othe exact string building we're doing so I'm a bit confused.

Correct.  I'm saying (to avoid duplicating logic, and also for clarity) simply use the existing object to accomplish the same thing you are accomplishing.

> Regarding the injected dependency - this will be done after the split.

Sounds good.

Reply:
I think this is a great thing to do together in Boston. it will be very educating to do this together.

Reply:
@orenbm @jonahx Let's shoot for working on UI and not authenticators together while we are in Boston

Reply:
NP, I added a ticket for this. we'll address it after the merge